### PR TITLE
Optimize composite agg with leading global ordinal value source

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
When queries are present in a search with a composite agg with a leading source that is of type `GlobalOrdinalValuesSource` there is an optimization we can do. In particular, once the composite queue is full, we know the range of ordinals we are interested in from the source. Thus, we can add a competitive iterator to the `LeafBucketCollector` that skips documents that are out of the competitive range.

This commit adds that optimization. In a dataset I have experimented with that has ~31M docs and a keyword with 200 different terms, I observed a 5x improvement in a simple search with a range query that matched ~28M docs and with `size = 5`.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java
Comment (dimitris-athanasiou) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java:160]: I know this is wrong. I had to add this for the competitive iterator to be picked up. I need some guidance on what this method should return.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java:55]: Do we actually need the boolean in that method? It looks like we're only ever doing something when it is true?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:55]: I believe this could cause issues with how TimeSeriesIndexSearcher is collecting mutiple segments in parallel. Or maybe `composite` cannot be used with `composite` anyway? Maybe @martijnvg has thoughts on this.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:249]: This feels more like a trace than a debug to me?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java:55]: The reason I added it is because we call this every time we add a slot to the queue but the first `size` times we don't want to update the bounds yet. I could change it so that we only call it when the queue is already full. But then we won't have a value for `highestCompetitiveValueGlobalOrd` until the queue fills up. I suppose that is not a problem.

Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java]: I have removed the unnecessary boolean and renamed the listener for clarity.

**Comment by dimitris-athanasiou** [comment]
@jpountz I have applied your suggestion regarding the score mode. However, I currently just check the leading source is a global ordinal to apply the optimization. For checking whether size is significantly smaller than cardinality of the source, I'll need some guidance.

As in the `CompositeAggregator` we do not have a `LeafReaderContext context` at hand, I am not sure how I could obtain the cardinality of the field.

**Comment by jpountz** [comment]
> For checking whether size is significantly smaller than cardinality of the source, I'll need some guidance.

I needed to look at the code as I was thinking so I pushed a commit that implements what I had in mind. I don't like that it's a bit sophisticated, but at the same time I wanted to have a way to disable the optimization in cases when the optimization might introduce some overhead while not helping, e.g. when `size` is greater than the cardinality of the field and there is a single source. I'm curious what you think of these heuristics, they're certainly not perfect.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:59]: The `composite` aggregation sometimes doesn't work with the `TimeSeriesIndexSearcher`.

In the cases that the `getLeafCollector(...)` returns `LeafBucketCollector.NO_OP_COLLECTOR` then search execution fails with a `collection_terminated_exception` error. In other cases composite does seem to work with `TimeSeriesIndexSearcher`. We should look into how to fix this. The problem is related why sample agg or the filter by filter optimisation doesn't work. 

However if we make this change then there is another reason why `composite` aggregation doesn't work and that is because `currentCompetitiveIterator` will likely refer to the wrong `CompetitiveIterator` instance. The `TimeSeriesIndexSearcher` creates `LeafWalker` (with `LeafBucketCollector` instance) instances for all segments before performing the search. This `LeafWalker` instances are used throughout the search and keep track aggregation state as the search executes on tsid by tsid basis. But then `currentCompetitiveIterator` will not be updated, since  all leaf buckets have already been created.

I don't think this problem should block this change. The `composite` agg is already not working correctly with `TimeSeriesIndexSearcher`. I think for now we should add validation that prevents the usage of `time_series` aggregation with the `composite` aggregation. (and also for the `sampler` aggregation)

Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:59]: > I don't think this problem should block this change. The composite agg is already not working correctly with TimeSeriesIndexSearcher. I think for now we should add validation that prevents the usage of time_series aggregation with the composite aggregation. (and also for the sampler aggregation)

Is that when the `time_series` agg is a parent of the composite or a child?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java:636]: This adds an addition in a hot loop. Maybe a better debug information would be to tell the number of segments on which `init()` was called, which would tell us whether dynamic pruning kicked in or not.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:97]: The check on the field type looks too fragile to me: if we were to introduced a new field type that also uses ordinals, then we'd probably miss updating this exclusion list. I would rather check if the index has terms  like what is done here: https://github.com/elastic/elasticsearch/pull/92060/files#diff-f61bc9e8ad11882e081009eb8e86945004041b3326497cec8672414645268dafR207-R208. Maybe add the IndexReader to this method signature so that it can check the content of the index and propagate it from `CompositeAggregator` which should have access to it via the `AggregationContext`.

**Review by jpountz — COMMENTED** [review]
The logic looks correct to me. One thing I'd like more covered in tests is the handling of the `after` key, and tests where there is a secondary source.

From this point I'll need folks from the Analytics team who are more familiar with the framework to help with the review from the perspective of doing things the right way for the agg framework since I've not coded against it in a while and may be a bit outdated.

**Comment by jpountz** [comment]
Something else that would be good would be to check if we can get this optimization to help with higher values of `size` like 50 rather than 5. We'd probably need a dataset with a higher cardinality than the one you tested with since the optimization would get disabled with size=50 and only 200 unique terms.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:59]: This is when `time_series` agg is a sub aggregator of `composite` aggregator. I don't think that `composite` aggregator allows `time_series` aggregator to be its parent aggregator.

We will address this in elastic/elasticsearch#92390.

**Comment by felixbarny** [comment]
@ruflin @jasonrhodes isn't the waffle map using composite aggregations? If so, this could be quite impactful for infra o11y. Could you help to quantify the impact?

**Comment by dimitris-athanasiou** [comment]
> Something else that would be good would be to check if we can get this optimization to help with higher values of size like 50 rather than 5. We'd probably need a dataset with a higher cardinality than the one you tested with since the optimization would get disabled with size=50 and only 200 unique terms.

I've tested with higher cardinality and size.

cardinality = 1000, #docs = ~31M

size = 50
main (avg) = 250 - 300 ms
opt (avg) = 130-160ms

size = 100
main (avg) = 230 - 250 ms
opt (avg) = 170ms-180ms

**Comment by dimitris-athanasiou** [comment]
@jpountz The tests I have added do test pagination. I have addressed your newest review comments. Could you please take another look?

**Comment by ruflin** [comment]
> @ruflin @jasonrhodes isn't the waffle map using composite aggregations? If so, this could be quite impactful for infra o11y. Could you help to quantify the impact?

On my end, I need to better understand what exactly the data + aggs look like that get improved. It is not clear to me what `GlobalOrdinalValuesSource` means for example. Any chance to provide a sample aggregation with a sample event as an example?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:114]: Getting an IOException here means that something bad happens, I would rather propagate it (possibly though an `UncheckedIOException` if we don't want to change method signatures)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:305]: nit: maybe also add the total number of segments seen, so that it's possible to know whether it was used on all segments or just a small subset?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:398]: maybe we should use the `docValuesFunc` that is configured on this agg?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:453]: When there is a single source, we should be able to bump this by 1 I think, since the after key is excluded? The implementation might be a bit tricky and require updating initLookup to record the first ordinal that compares strictly greater than `afterKey`, which might be the same as `lowOrd` when the `afterKey` does not exist in the index. We could look into it in a follow-up PR and just leave a comment for now.

**Review by jpountz — COMMENTED** [review]
One thing that is missing is a check for `fieldType.isIndexed()` similar to `ValuesSourceConfig#alignesWithSearchIndex` to make sure that we are not fetching data from the index while actually operating on a runtime field that shadows an indexed field. Otherwise the change looks good, I left some minor comments inline.

**Comment by jpountz** [comment]
> I've tested with higher cardinality and size.

Thanks for running these benchmarks! I was hoping for a bit more, but 2x with size=50 is quite fantastic already!

> On my end, I need to better understand what exactly the data + aggs look like that get improved

To my understanding, here are the conditions that need to be met:
 - The leading source of the composite aggregation is of type `terms` and runs on an indexed `keyword` field.
 - `missing_bucket` is not configured.
 - The cardinality of the field is much higher than the `size` parameter (ie. retrieving a single page of data retrieves a small subset of the data). The optimization disables itself if the cardinality is not at least `8 * size` if there is a single source, or `2.6 * size` if there are multiple sources.
 - A single page of data retrieves a somewhat contained number of unique values. The optimization disables itself if `size > 100` for a single source or `size > 300` for multiple sources.
 - How well the optimization works also highly depends on how values are spread across the index. Having values uniformly spread will help the optimization kick in earlier by more quickly converging towards the top `size` buckets for the shard. In a follow-up, we could look into further specializing this optimization for e.g. TSDB where the index sort would create clusters of equal values.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:453]: Bumping by 1 when there is a single source makes sense to me. This should be beneficial on its own. I do not fully understand the benefit of also recording the first ordinal > `afterKey`. Even though the competitive iterator might have a wider range on the low side, the actual query should skip non-matching docs anyway, right?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:305]: Effectively `num_global_ordinal_dynamic_pruning_initialized` is the number of segments. We create one `LeafBucketCollector` per segment, thus potentially up to one competitive iterator per segment. And we count the number of times it actually gets initialized. Perhaps I should rename this to something like `num_segments_global_ordinal_dynamic_pruning_used`?

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/bwc

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:312]: Sorry I meant to track the total number of segments seen by this aggregator, including those where init() is never called with something like `segments_collected` as a key (consistent with FilterByFilterAggregator). As far as naming is concerned, it looks like a few other aggregators adopted the naming convention of prefixing segment counts with `segments_` (e.g. FilterByFilterAggregator), so `segments_dynamic_pruning_used`? (I made it a bit shorted by removing references to global ords since the terms source only works with global ords anyway?)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:456]: There is a corner case when the after key does not exist in a shard. For instance say a shard has two values `a` and `c`, and the `after_key` is `b`. The `after` global ord will be set to the global ordinal of `c` in that case, so it shouldn't be bumped here.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:312]: I see. OK, will do that.

As far as naming is concerned, I thought having `global_ordinals` in there would distinguish this debug info from other debug info that could potentially come from a nested aggregation. But I'll do the changes anyway and if we have a name clash in the future we can resolve it then.

Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:312]: Another idea would be to prefix with `sources.{terms_source_name}.segments_dynamic_pruning_used` and `sources.{terms_source_name}.segments_collected`. What do you think?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:456]: I would expect us to have a check on whether there is a single source (we can't skip the ord of the after key if there are other sources, e.g. say your after key is `[ "c", "d" ]`, then `["c", "e"]` also compares greater than the after key, while the leading source has a value equal to the same value as in the after key). But indeed, something like that. Let's make sure we have good tests for this case to make sure we don't skip more than would be correct?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:312]: > distinguish this debug info from other debug info that could potentially come from a nested aggregation

Inner aggregations should put their debug info in their own namespace, so we don't need to worry about naming collisions.

Adding the name of the source to the debug info makes sense to me. Tagging @martijnvg in case he has thoughts on this.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:456]: Right yes. OK, I think I prefer to add a TODO about this additional optimization for now and leave it for a follow up PR.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java]: Agreeing with this statement here.

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/part-3

**Review by martijnvg — COMMENTED** [review]
In general this change looks good to me too.
Thanks for improving `composite` aggregation @dimitris-athanasiou!

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:61]: Fair point. I'll remove it.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:464]: I will raise an issue as soon as this is merged.

**Review by jpountz — APPROVED** [review]
LGTM, thanks!

**Comment by martijnvg** [comment]
LGTN

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
Comment (dimitris-athanasiou) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java:464]: I raised this in https://github.com/elastic/elasticsearch/issues/92505

**Comment by jpountz** [comment]
> I've tested with higher cardinality and size.

Out of curiosity, what did your search request look like? Was the query a range query by any chance?

**Comment by jpountz** [comment]
(I'm speculating that maybe we did not see a greater speedup because evaluating the query has a non negligible cost, and we might see greater speedups when the query is cheaper.)

**Comment by dimitris-athanasiou** [comment]
> Out of curiosity, what did your search request look like? Was the query a range query by any chance?

Yes it was. It was this:

```
{
    "size":0,
    "query": {
        "bool": {
            "must": {
                "range": {
                    "responsetime": {
                        "gt": 100
                    }
                }
            }
        }
    },
    "aggs": {
        "buckets": {
            "composite": {
                "size": 50,
                "sources": [
                    {"airline": {"terms": {"field": "airline"}}}
                ]
            }
        }
    }
}
```

**Comment by jpountz** [comment]
Thank you for sharing the query. For the record, here's what is in my mind: one issue we have with top level range queries is that evaluating hits requires (essentially) populating a bitset of matching doc IDs, so that runs in linear time with the number of matches.

So as we are improving the `composite` aggregation to only look at a small subset of matches with this PR, it's quite likely that the bottleneck is not running the aggregation, but evaluating the set of matching documents, hence why the speedup is not so impressive.

There have been separate discussions about enabling index sorting on `@timestamp` on Logging datasets, so that evaluating the set of matching documents boils down to identifying the first and last matching documents (~ `O(log num_matches)` instead of ~ `O(num_matches)`). If/when we do this, it's possible that we'd see an even greater speedup on composite aggregations when the query is a range filter on `@timestamp`.

If your dataset allows, it would be interesting to see what the speedup looks like when the query is a `term` query on a `keyword` field that matches lots of documents.

**Pull request merged**
